vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs

```diff
@@ -0,0 +1,476 @@
+        |> List.choose id
+        |> longIdentToArray
+
+    let getOpenDeclarations (ast: ParsedInput option) (entities: RawEntity list option) 
+                            (qualifyOpenDeclarations: Line -> EndColumn -> Idents -> Async<RawOpenDeclaration list>) = async {
+        match ast, entities with
+        | Some (ParsedInput.ImplFile (ParsedImplFileInput(_, _, _, _, _, modules, _))), Some entities ->
+            let autoOpenModulesAndActivePatterns = 
+                getAutoOpenModules entities @ getActivePatterns entities
+            //debug "All AutoOpen modules: %A" autoOpenModules
+             
+            let rec walkModuleOrNamespace acc (decls, moduleRange) = async {
+                let rec loop acc exprs = async {
+                    match exprs with
+                    | [] -> return acc
+                    | SynModuleDecl.NestedModule (_, _, nestedModuleDecls, _, nestedModuleRange) :: rest -> 
+                        let! acc' = walkModuleOrNamespace acc (nestedModuleDecls, nestedModuleRange)
+                        return! loop acc' rest
+                    | SynModuleDecl.Open (LongIdentWithDots(longIdent, _), openStatementRange) :: rest ->
+                        let identArray = processOpenDeclaration longIdent
+                        let! rawOpenDeclarations =  
+                            identArray |> qualifyOpenDeclarations openStatementRange.StartLine openStatementRange.EndColumn
+
+                        for openDecl in rawOpenDeclarations do
+                            Debug.Assert (openDecl.Idents |> Array.endsWith identArray, 
+                                            sprintf "%A must be suffix for %A" identArray openDecl.Idents)
+                            for ident in openDecl.Idents do
+                                Debug.Assert (IdentifierUtils.isIdentifier ident,
+                                              sprintf "%s as part of %A must be an identifier" ident openDecl.Idents)
+
+                        (* The idea that each open declaration can actually open itself and all direct AutoOpen modules,
+                            children AutoOpen modules and so on until a non-AutoOpen module is met.
+                            Example:
+                               
+                            module M =
+                                [<AutoOpen>]                                  
+                                module A1 =
+                                    [<AutoOpen>] 
+                                    module A2 =
+                                        module A3 = 
+                                            [<AutoOpen>] 
+                                            module A4 = ...
+                                     
+                            // this declaration actually open M, M.A1, M.A1.A2, but NOT M.A1.A2.A3 or M.A1.A2.A3.A4
+                            open M
+                        *)
+
+                        let rec loop' acc maxLength =
```

Comment:
I'm concerned how hard this sort of code is to understand (``let rec loop'`` followed further down by another ``let rec loop``).  The iteration patterns can almost always be abstracted.  At the moment it's really hard to follow,